src/ripple/app/reporting/P2pProxy.h

```diff
@@ -0,0 +1,115 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2020 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_CORE_TXPROXY_H_INCLUDED
+#define RIPPLE_CORE_TXPROXY_H_INCLUDED
+
+#include <ripple/app/main/Application.h>
+#include <ripple/rpc/Context.h>
+#include <ripple/rpc/impl/Handler.h>
+
+#include <boost/beast/websocket.hpp>
+
+#include "org/xrpl/rpc/v1/xrp_ledger.grpc.pb.h"
+#include <grpcpp/grpcpp.h>
+
+namespace ripple {
+/// Forward a JSON request to a p2p node and return the response
+/// @param context context of the request
+/// @return response from p2p node
+Json::Value
+forwardToP2p(RPC::JsonContext& context);
+
+/// Whether a request should be forwarded, based on request parameters
+/// @param context context of the request
+/// @return true if should be forwarded
+bool
+shouldForwardToP2p(RPC::JsonContext& context);
+
+template <class T>
+struct SpecifiesLedger
+{
+    // List out all GRPC request types that specify a ledger
+    // Note, GetAccountTransactionHistory specifies a ledger, but
+    // GetAccountTransationHistory only ever returns validated data, so
+    // GetAccountTransactionHistory will never be forwarded
+    static bool const value =
+        std::is_same<T, org::xrpl::rpc::v1::GetAccountInfoRequest>::value ||
+        std::is_same<T, org::xrpl::rpc::v1::GetLedgerRequest>::value ||
+        std::is_same<T, org::xrpl::rpc::v1::GetLedgerDataRequest>::value;
+};
+
+/// Whether a request needs the current or closed ledger
+/// @param context context of the request
+/// @return true if the request needs the current or closed ledger
+template <
+    class Request,
+    typename std::enable_if<!SpecifiesLedger<Request>::value, Request>::type* =
+        nullptr>
```

Comment:
In C++17 this can be simplified to:
```c++
template <
    class Request,
    std::enable_if_t<!SpecifiesLedger<Request>::value>* =  nullptr>
```
(two places)

Reply:
We can also remove one of the overloads, get rid of the `enable_if` and rewrite this with `if constexpr`. Something like:
```
template <class Request>
bool
needCurrentOrClosed(RPC::GRPCContext<Request>& context)
{
    if constexpr (SpecifiesLedger<Request>::value)
    {
        if (context.params.ledger().ledger_case() ==
            org::xrpl::rpc::v1::LedgerSpecifier::LedgerCase::kShortcut)
        {
            if (context.params.ledger().shortcut() !=
                    org::xrpl::rpc::v1::LedgerSpecifier::SHORTCUT_VALIDATED &&
                context.params.ledger().shortcut() !=
                    org::xrpl::rpc::v1::LedgerSpecifier::SHORTCUT_UNSPECIFIED)
                return true;
        }
    }
    return false;
}
```

Reply:
I followed the suggestion of @seelabs: used `if constexpr` and removed the `SpecifiesLedger` struct entirely. I also realized this was missing a few request types, so I added those. Also added unit tests for this function (and changed the function signature to make it easier to unit test)